### PR TITLE
Issue 650 - Updated LocationLevel getOne to handle custom units

### LINKED ISSUE #650: Location Levels should allow arbitrary units
#596 indicates that the current location levels endpoints assumes users will only required "SI" units. This is an incorrect assumptions.
Location levels can be any unit and the user should be allowed to specify them.

### PR DESCRIPTION
**Pull request by zack-rma**
Fixes #650 - Updated LocationLevel getOne to handle custom units and unit systems

### CODE REVIEW

**Comment by zack-rma** [comment]
I also discovered that the effective date query parameter for getOne is used as the end date in the PL/SQL call. Effectively, any date provided that is after the effective date of the target locationLevel will work to find that locationLevel. This also results in the returned effective date being set to the provided value, not the DB value.

Not sure about the impact of this, maybe it should get its own issue? @MikeNeilson

**Review by MikeNeilson — DISMISSED** [review]

**Comment by MikeNeilson** [comment]
> I also discovered that the effective date query parameter for getOne is used as the end date in the PL/SQL call. Effectively, any date provided that is after the effective date of the target locationLevel will work to find that locationLevel. This also results in the returned effective date being set to the provided value, not the DB value.
> 
> Not sure about the impact of this, maybe it should get its own issue? @MikeNeilson

Sorry, I've re-read this several times and I'm still having difficulty following the issue; I mean it seems like an issue but it's not mapping in my head. Any way to make a simple diagram, flow chart, just plain example data where the issue would be more obvious to myself and others.?

**Comment by adamkorynta** [comment]
If I'm following correctly: lets say I pass in today's date, it'll give me back the loc level value from the last effective date (let's say last year), but the DTO returns today's date as the effective date.

**Comment by MikeNeilson** [comment]
Ah. I see, yeah the correct effective date should be what's in the DTO.

**Comment by zack-rma** [comment]
Rebased to fix build errors.

**Comment by zack-rma** [comment]
Rebased to fix build issue

**Comment by zack-rma** [comment]
Build errors appear to be unrelated to these changes

File: cwms-data-api/src/main/java/cwms/cda/data/dao/LocationLevelsDaoImpl.java [hunk 1]
```diff
@@ -330,6 +331,11 @@ public LocationLevel retrieveLocationLevel(String locationLevelName, String pUni
         return connectionResult(dsl, c -> {
             String units = pUnits;
             Configuration configuration = getDslContext(c, officeId).configuration();
+            if (units != null && (units.equalsIgnoreCase("SI")
+                    || units.equalsIgnoreCase("EN"))) {
+                units = CWMS_UTIL_PACKAGE.call_GET_DEFAULT_UNITS(configuration,
+                        locationLevelName.split("\\.")[1], units);
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dao/LocationLevelsDaoImpl.java]: I don't see validation on the location level id prior to this, the incoming string might not have any periods and would then throw and ArrayIndexOutOfBoundsException

**Review by adamkorynta — CHANGES_REQUESTED** [review]

File: cwms-data-api/src/main/java/cwms/cda/data/dao/LocationLevelsDaoImpl.java (hunk 1 above)
Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/data/dao/LocationLevelsDaoImpl.java]: Added a format check before calling methods using the split string

Comment (adamkorynta) [reply cwms-data-api/src/main/java/cwms/cda/data/dao/LocationLevelsDaoImpl.java]: you're splitting the string twice (and not using a compiled Pattern) - should reuse the split for efficiency

Comment (zack-rma) [reply cwms-data-api/src/main/java/cwms/cda/data/dao/LocationLevelsDaoImpl.java]: Added a pattern and cleaned up usage

**Review by adamkorynta — APPROVED** [review]

**Pull request merged**